File: frontend/src/types/purchase-order.ts
export interface POLine {
  id: string;
  po_id: string;
  product_id: string | null;
  product_name: string | null;
  description: string;
  quantity_ordered: number;
  quantity_received: number;
  unit_cost: number;
  line_total: number;
  sort_order: number;
  created_at: string;
}

export interface POLineCreate {
  product_id?: string | null;
  description: string;
  quantity_ordered: number;
  unit_cost: number;
  sort_order?: number;
}

export interface PurchaseOrder {
  id: string;
  company_id: string;
  number: string;
  vendor_id: string;
  vendor_name: string | null;
  status: string;
  order_date: string;
  expected_date: string | null;
  shipping_address: string | null;
  subtotal: number;
  tax_amount: number;
  total: number;
  notes: string | null;
  sent_at: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  modified_at: string | null;
  lines: POLine[];
}

export interface PurchaseOrderListItem {
  id: string;
  number: string;
  vendor_id: string;
  vendor_name: string | null;
  status: string;
  order_date: string;
  expected_date: string | null;
  subtotal: number;
  tax_amount: number;
  total: number;
  created_at: string;
}

export interface PurchaseOrderCreate {
  vendor_id: string;
  order_date?: string;
  expected_date?: string;
  shipping_address?: string;
  tax_amount?: number;
  notes?: string;
  lines: POLineCreate[];
}

export interface PurchaseOrderUpdate {
  vendor_id?: string;
  order_date?: string;
  expected_date?: string;
  shipping_address?: string;
  tax_amount?: number;
  notes?: string;
  lines?: POLineCreate[];
}

export interface PaginatedPurchaseOrders {
  items: PurchaseOrderListItem[];
  total: number;
  page: number;
  per_page: number;
}

export interface POStats {
  total_pos: number;
  draft: number;
  sent: number;
  partial: number;
  received: number;
  closed: number;
}

export interface ReceiveLineItem {
  po_line_id: string;
  quantity_received: number;
}

File: backend/app/schemas/purchase_order.py
"""Pydantic schemas for Purchase Orders."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, field_validator


# ---------------------------------------------------------------------------
# Lines
# ---------------------------------------------------------------------------


class POLineCreate(BaseModel):
    product_id: str | None = None
    description: str
    quantity_ordered: Decimal
    unit_cost: Decimal
    sort_order: int = 0


class POLineUpdate(BaseModel):
    id: str | None = None  # existing line id (None = new line)
    product_id: str | None = None
    description: str | None = None
    quantity_ordered: Decimal | None = None
    unit_cost: Decimal | None = None
    sort_order: int | None = None


class POLineResponse(BaseModel):
    id: str
    po_id: str
    product_id: str | None = None
    product_name: str | None = None
    description: str
    quantity_ordered: Decimal
    quantity_received: Decimal
    unit_cost: Decimal
    line_total: Decimal
    sort_order: int
    created_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Receiving
# ---------------------------------------------------------------------------


class ReceiveLineItem(BaseModel):
    po_line_id: str
    quantity_received: Decimal


class ReceivePayload(BaseModel):
    lines: list[ReceiveLineItem]


# ---------------------------------------------------------------------------
# Purchase Order
# ---------------------------------------------------------------------------


class PurchaseOrderCreate(BaseModel):
    vendor_id: str
    order_date: str | None = None  # ISO date string, defaults to today
    expected_date: str | None = None
    shipping_address: str | None = None
    tax_amount: Decimal = Decimal("0.00")
    notes: str | None = None
    lines: list[POLineCreate] = []

    @field_validator("*", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        if isinstance(v, str) and v.strip() == "":
            return None
        return v


class PurchaseOrderUpdate(BaseModel):
    vendor_id: str | None = None
    order_date: str | None = None
    expected_date: str | None = None
    shipping_address: str | None = None
    tax_amount: Decimal | None = None
    notes: str | None = None
    lines: list[POLineUpdate] | None = None  # if provided, replaces all lines

    @field_validator("*", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: object) -> object:
        if isinstance(v, str) and v.strip() == "":
            return None
        return v


class PurchaseOrderResponse(BaseModel):
    id: str
    company_id: str
    number: str
    vendor_id: str
    vendor_name: str | None = None
    status: str
    order_date: datetime
    expected_date: datetime | None = None
    shipping_address: str | None = None
    subtotal: Decimal
    tax_amount: Decimal
    total: Decimal
    notes: str | None = None
    sent_at: datetime | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    modified_at: datetime | None = None
    lines: list[POLineResponse] = []

    model_config = {"from_attributes": True}


class PurchaseOrderListItem(BaseModel):
    id: str
    number: str
    vendor_id: str
    vendor_name: str | None = None
    status: str
    order_date: datetime
    expected_date: datetime | None = None
    subtotal: Decimal
    tax_amount: Decimal
    total: Decimal
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedPurchaseOrders(BaseModel):
    items: list[PurchaseOrderListItem]
    total: int
    page: int
    per_page: int


class POStats(BaseModel):
    total_pos: int
    draft: int
    sent: int
    partial: int
    received: int
    closed: int
